libinteract/libinteract.py

```diff
@@ -867,6 +868,144 @@ def calc_cg_fullmatrix(identifiers, idxs, percmat, perco):
+                                         chain1 == cols[4])
+        # Combine the vectors
+        logical_vector = np.logical_or(logical_vector1, logical_vector2)
+    # Filter rows
+    filtered_rows = rows[logical_vector]
+    # Warn if no contacts found and return None
+    if filtered_rows.shape[0] == 0:
+        if chain1 == chain2:
+            log.warning(f"No intrachain contacts found in chain {chain1}")
+        else:
+            warnstr = f"No interchain contacts found between chains " \
+                      f"{chain1} and {chain2}"
+            log.warning(warnstr)
+        return None
+    else:
+        return filtered_rows
+
+
+def create_table_dict(table):
+    """Takes in a single table (list of tuples) and returns a dictionary
+    of tables (arrays). Each key in this dictionary represent whether the 
+    table contains all/intrachain/interchain contacts.
+    """
+
+    # Convert to array and keep transpose for selection
+    table_rows = np.array(table)
+    table_cols = table_rows.T
+    # Initialize output dictionary of tables
+    table_dict = {"all": table_rows}
+    # Find unique chains in the table
+    chains = np.unique(np.concatenate((table_cols[0], table_cols[4])))
+    # If multiple chains are present, split the contacts by chain
+    if len(chains) > 1:
+	# For each chain, add the nodes that are only in contact with the same chain
+        for chain in chains:
+            filtered_rows_intra = filter_by_chain(chain, chain,
+                                                  table_rows)
+            # Check rows exist
+            if filtered_rows_intra is not None:
+                table_dict[chain] = filtered_rows_intra
+        # Create vector of all nodes that are in contact with different chains
+        logical_vector = table_cols[0] != table_cols[4]
+        # Warn if no interchain contacts
+        if logical_vector.sum() == 0:
+            log.warning("No interchain contacts found")
+        else:
+            # For all combinations of different chains, find nodes that are in
+            # contact with different chains
+            for chain1, chain2 in itertools.combinations(chains, 2):
+                filtered_rows_inter = filter_by_chain(chain1, chain2,
+                                                       table_rows)
+                # Check rows exist
+                if filtered_rows_inter is not None:
+                    name = tuple(sorted([chain1, chain2]))
+                    table_dict[name] = filtered_rows_inter
+    return table_dict
+
+def create_matrix_dict(fullmatrix, table_dict, pdb):
+    """Takes in the full matrix of persistence values and a dictionary of 
+    tables where each key represents all/intrachain/interchain contacts 
+    and returns a dictionary of matrices for each key.
+    """
+
+    # Initialize output dictionary of matrices
+    mat_dict = {"all": fullmatrix}
+    # Map each residue id to a position in the matrix
+    resids = pdb.residues.resids
```

Comment:
this should use `pdb.residues.resindices` instead of `resids`. `resids` contains the residue name according to the definition in the PDB, which means that residues in two chains could have the same `resid` but be distinct. `resindices` contains residue indices that are unique for every residue

Reply:
I couldn't make this change since I use the columns of the table to choose the matrix index, and the `resindices` don't correspond to anything in the table. I changed it so that I use the chain id + the res num (e.g. 'A4') since both these pieces of information are already in the table, and I assume even if two residues have the same res number, they must have different chain ids e.g. A4 and B4 so there won't be any conflicts. I hope this is okay.